setup.php

```diff
@@ -244,12 +251,76 @@ function recursive_writable_directory_test($dir)
 
 // Include the sqlconf file if it exists yet.
 $config = 0;
-if (file_exists($OE_SITE_DIR)) {
-    include_once($installer->conffile);
+if (file_exists($installer->conffile)) {
+    require_once($installer->conffile);
 } elseif ($state > 3) {
   // State 3 should have created the site directory if it is missing.
+    SessionUtil::setupScriptSessionStart();
+    SessionUtil::setupScriptSessionCookieDestroy();
     die("Internal error, site directory is missing.");
 }
+
+
+// This will effectively only allow entry into the setup.php script at the first step and will bar entry to the
+//  script if openemr has already been installed.
+//  2 mechanisms are also in place to ensure go through script in chronological order
+//   - a session variable tracks the next state expected and kills script is not correct
+//   - only the next state csrf is handed out in the form
+if (empty($state)) {
+    // ensure not already installed
+    if (!empty($config)) {
+        SessionUtil::setupScriptSessionStart();
+        SessionUtil::setupScriptSessionCookieDestroy();
+        error_log("OpenEMR has already been installed. If you wish to force re-installation, then edit " . errorLogEscape($installer->conffile) . " (change the 'config' variable to 0), and re-run the setup.php script.");
+        die("OpenEMR has already been installed. If you wish to force re-installation, see log for details.<br />\n");
+    }
+    // set up new blank session and csrf mechanism
+    SessionUtil::setupScriptSessionStart();
+    session_regenerate_id(true);
+    $_SESSION = [];
+    CsrfUtils::setupCsrfKey();
+} else {
+    // start up session and check csrf
+    SessionUtil::setupScriptSessionStart();
+    $state = (int)$state;
+    $verifyCsrf = false;
+    if (($state > 0) && ($state < 8)) {
```

Comment:
Anyway we can put these numbers as constants somewhere?  Makes it hard to follow the logic of what is going on with 0, 8, 3, etc.

Reply:
i could, but then I'd begin to consider Seppuku :)
i'll do the next best thing, though, and put some comments about what each one is

Reply:
lol, well let's not get hasty haha.

Reply:
luckily no sword handy, but there is always amazon prime :)

Reply:
added more description of the state(s) to the script header comments